pedsnetdcc/age_transform.py

```diff
@@ -0,0 +1,142 @@
+$$""",
+"""comment on function month_after_last_month_of_interval(timestamp, timestamp) is
+'Return the timestamp of the month AFTER the last month of the interval between two timestamps'""",
+
+"""create or replace function days_in_last_month_of_interval(timestamp, timestamp)
+     returns double precision strict immutable language sql as $$
+       select extract(days from month_after_last_month_of_interval($1, $2) - last_month_of_interval($1, $2))
+$$""",
+"""comment on function days_in_last_month_of_interval(timestamp, timestamp) is
+'Return the number of days in the last month of the interval between two timestamps'""",
+
+"""create or replace function months_in_interval(timestamp, timestamp)
+ returns double precision strict immutable language sql as $$
+  select extract(years from age($2, $1)) * 12 + extract(months from age($2, $1)) + extract(days from age($2, $1))/days_in_last_month_of_interval($1, $2)
+$$""",
+"""comment on function months_in_interval(timestamp, timestamp) is
+   'Return the number of months (double precision) between two timestamps.
+    The number of years/months/days is computed by PostgreSQL''s
+    extract/date_part function.  The fractional months value is
+    computed by dividing the extracted number of days by the total
+    number of days in the last month overlapping the interval; note
+    that this is not a calendar month but, say, the number of days
+    between Feb 2, 2001 and Mar 2, 2001.  You should be able to obtain
+    the original timestamp from the resulting value, albeit with great
+    difficulty.'"""
+)
+
+class AgeTransform(Transform):
+    # Caller may override `columns`
+    columns = [
+        ('condition_occurrence', 'condition_start_time'),
+        ('death', 'death_time'),
+        ('drug_exposure', 'drug_exposure_start_time'),
+        ('measurement', 'measurement_time'),
+        ('measurement', 'measurement_result_time'),
+        ('procedure_occurrence', 'procedure_time'),
+        ('visit_occurrence', 'visit_start_time'),
+        ('observation', 'observation_time')
+    ]
+    AGE_COLUMN_TYPE = 'float'
+
+    @classmethod
+    def is_age_column(cls, column_name, table_name):
+        # Almost makes me want to change the data structure
+        for age_table, age_col in cls.columns:
+            if age_table == table_name and age_col == column_name:
+                return True
+        return False
+
+    @classmethod
+    def pre_transform(cls, dburi, search_path):
+        """Define PL/SQL functions needed for the age transform
+        See also Transform.pre_transform.
+        """
+        with psycopg2.connect(dburi) as conn:
+            with conn.cursor() as cursor:
+
+                if search_path:
+                    # SQL injection vulnerability?
+                    cursor.execute(
+                        'SET search_path TO {0}'.format(search_path))
+
+                for stmt in AGE_FUNCTIONS_SQL:
+                    cursor.execute(stmt)
+
```

Comment:
Need to add a `conn.close()` because the context manager of psycopg2 connections does not actually close the connection, it just commits the transaction. I know. Weird. http://initd.org/psycopg/docs/usage.html#with-statement